src/beacon.js

```diff
@@ -94,39 +124,53 @@ export default function Emitter(endpoint, context, options) {
         let editedMessage = message;
         editedMessage = message.payload || {};
         editedMessage.type = message.type;
-        if (!send(endpoint, JSON.stringify(editedMessage), sync) && retries) {
-          retries -= 1;
-          messages.push(message);
-          console.error('Evolv: Unable to send event beacon');
-        }
+        send(endpoint, JSON.stringify(editedMessage), sync);
       });
     } else {
       // eslint-disable-next-line no-constant-condition
       while (true) {
-        const smallBatch = batch.slice(0, BATCH_SIZE);
-        if (smallBatch.length === 0) {
+        if (batch.length === 0) {
           break;
         }
 
-        if (!send(endpoint, JSON.stringify(wrapMessages(smallBatch)), sync) && retries) {
-          retries -= 1;
-          messages = smallBatch;
-          console.error('Evolv: Unable to send analytics beacon');
-          break;
+        let reducedBatchSize = 0;
+        let charCount = 0;
+        for (let i = 0; i < (batch.length && BATCH_SIZE); i++) {
+          charCount += encodeURIComponent(JSON.stringify(batch[i])).length;
+
+          if (charCount > MAX_MESSAGE_SIZE) {
+            break;
+          }
+
+          reducedBatchSize = i + 1;
         }
 
-        batch = batch.slice(BATCH_SIZE)
+        const smallBatch = batch.slice(0, reducedBatchSize);
+
+        // smallBatch would be 0 if the first message was too big
+        // just grab the first message, and force send it with failover
+        if (smallBatch.length === 0) {
+          send(endpoint, JSON.stringify(wrapMessages([batch[0]])), sync, true);
+          batch = batch.slice(1);
+          continue;
+        }
+
+        send(endpoint, JSON.stringify(wrapMessages(smallBatch)), sync);
+        batch = batch.slice(reducedBatchSize);
       }
     }
 
     if (messages.length) {
-      timer = setTimeout(function() {transmit(retries)}, DELAY);
+      timer = setTimeout(function() {transmit()}, DELAY);
     }
   }
 
   if (typeof window !== 'undefined') {
-    window.addEventListener('unload', transmit);
-    window.addEventListener('beforeunload', transmit);
+    window.addEventListener("visibilitychange", function transmitData() {
```

Comment:
For clarity perhaps:
```suggestion
/** @see {@link https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event#sending_end-of-session_analytics_on_transitioning_to_hidden} */
    window.addEventListener("visibilitychange", function transmitData() {
```

Reply:
Added